Home_Work_26/VectorsFunctions.cpp

```diff
@@ -0,0 +1,221 @@
+	float result_EndY_axis = this->getEndYValue() - secondVector.getEndYValue();
+	float result_StartX_axis = this->getStartXValue() - secondVector.getStartXValue();
+	float result_StartY_axis = this->getStartYValue() - secondVector.getStartYValue();
+
+	return *(new Vector2d(result_EndX_axis, result_EndY_axis, result_StartX_axis, result_StartY_axis));
+}
+
+float Vector2d::operator()()
+{
+	float xLengthPowerofTwo = std::pow(this->getEndXValue() - this->getStartXValue(), 2);
+	float yLengthPowerofTwo = std::pow(this->getEndYValue() - this->getStartYValue(), 2);
+
+	float vectorLegth = std::sqrt(xLengthPowerofTwo + yLengthPowerofTwo);
+
+	return vectorLegth;
+}
+
+float& Vector2d::operator[](int iterator)
+{
+	MY_ASSERT(iterator>=0 && iterator<4, "Wrong index")
+	if (iterator == 0)
+	{
+		return this->m_EndX_axis;
+	}
+
+	else if (iterator == 1)
+	{
+		return this->m_EndY_axis;
+	}
+	else if (iterator == 2)
+	{
+		return this->m_StartX_axis;
+	}
+	else if (iterator == 3)
+	{
+		return this->m_StartY_axis;
+	}
+}
+
+std::ostream& operator<<(std::ostream& out, Vector2d& vector)
+{
+	out << "Vector paramaters: {" << vector.m_EndX_axis << ", " << vector.m_EndY_axis << ", " << vector.m_StartX_axis << ", " << vector.m_StartY_axis <<  "}";
+	return out;
+}
+
+
+void operator >>(std::istream& in, Vector2d& vector)
+{
+	in >> vector.m_EndX_axis;
+	in >> vector.m_EndY_axis;
+	in >> vector.m_StartX_axis;
+	in >> vector.m_StartY_axis;
+}
+
+
+float Vector2d::dotProduct(Vector2d& other) const
+{
+	float firstStartXParameter = this->getEndXValue() - this->getStartXValue();
+	float firstStartYParameter = this->getEndYValue() - this->getStartYValue();
+	float secondStartXParameter = other.getEndXValue() - other.getStartXValue();
+	float secondStartYParameter = other.getEndYValue() - other.getStartYValue();
+
+	return firstStartXParameter * secondStartXParameter + firstStartYParameter * secondStartYParameter;
+}
+
+Vector2d Vector2d::negate() const 
+{
+	Vector2d* invertedVector = new Vector2d(-this->getEndXValue(), -this->getEndYValue(), -this->getStartXValue(), -this->getStartYValue());
+	return  *invertedVector;
+}
+
+float Vector2d::getLength() const
+{
+	float xLengthPowerofTwo = std::pow(this->getEndXValue() - this->getStartXValue(), 2);
+	float yLengthPowerofTwo = std::pow(this->getEndYValue() - this->getStartYValue(), 2);
+
+	float vectorLegth = std::sqrt(xLengthPowerofTwo + yLengthPowerofTwo);
+
+	return vectorLegth;
+}
+
+VectorRelativeState Vector2d::getRelativeState(Vector2d& other) const
+{
+	float dotProduction = this->dotProduct(other);
+	float firstVecLength = this->getLength();
+	float secondVecLength = other.getLength();
+	float cosTeta = dotProduction / (firstVecLength * secondVecLength);
+	float arccosTeatRad = std::acos(dotProduction / (firstVecLength * secondVecLength));
+	float arccosTeatDeg = arccosTeatRad * (180 / pi);
+
+	bool endIdentical = (this->getEndXValue() == other.getEndXValue() && this->getEndYValue() == other.getEndYValue());
```

Comment:
comparing float's directly is not that simple. They must be compared up to some precision, see https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon